core/src/main/scala/chisel3/experimental/HasAutoTypename.scala

```diff
@@ -0,0 +1,53 @@
+package chisel3.experimental
+
+import chisel3.{Data, Record}
+import chisel3.internal.{sanitize}
+
+/** Trait for [[Record]]s that signals the compiler plugin to generate a typeName for the
+  * inheriting [[Record]] based on the parameter values provided to its constructor.
+  *
+  * @example Consider a [[Bundle]] which manually implements `typeName`:
+  * {{{
+  * class MyBundle(param1: Int, param2: Int)(gen: Data) extends Bundle {
+  *   val foo = UInt(param1.W)
+  *   val bar = UInt(param2.W)
+  *   val data = gen
+  *   override def typeName = s"MyBundle_\${param1}_\${param2}_\${gen.typeName}"
+  * }
+  *
+  * (new MyBundle(3, 4)(SInt(3.W))).typeName // "MyBundle_3_4_SInt3"
+  * (new MyBundle(1, 32)(Bool())).typeName   // "MyBundle_1_32_Bool"
+  * }}}
+  *
+  * An identical `typeName` implementation can be generated and provided with `HasAutoTypename`, making
+  * the manual implementation unnecessary:
+  * {{{
+  * class MyBundle(param1: Int, param2: Int)(gen: Data) extends Bundle with HasAutoTypename {
+  *   val foo = UInt(param1.W)
+  *   val bar = UInt(param2.W)
+  *   val data = gen
+  * }
+  *
+  * (new MyBundle(3, 4)(SInt(3.W))).typeName // "MyBundle_3_4_SInt3"
+  * (new MyBundle(1, 32)(Bool())).typeName   // "MyBundle_1_32_Bool"
+  * }}}
+  */
+trait HasAutoTypename {
+  this: Record =>
+
+  /** Auto generate a type name for this Bundle using the bundle arguments supplied by the compiler plugin.
+    */
+  override def typeName: String = autoTypeName(getClass.getSimpleName, _typeNameConParams)
```

Comment:
We should probably think about what the behavior should be for anonymous classes, eg. what if someone writes:
```scala
class MyModule(n: Int) extends Module {
  val io = IO(new Bundle with HasAutoTypename {
    val foo = Input(UInt(n.W))
  })
}
```
I think you will get a name that is something like `_packageMyModulee0e14f5_8` which is a pretty terrible name indeed.

I think there's a compelling argument that you should not be able to mix HasAutoTypename into anonymous classes since it's going to be a disaster, and arguably not any inner class (inner classes have their outer object as the first argument, that's where `packageMyModuleee0e145f` comes from, that's the `.toString` of the instance of `MyModule` above).

Also `.getSimpleName` throws an exception in some circumstances on Java8 which is still supported by Chisel, see https://github.com/scala/bug/issues/2034 and this is why we have more complicated logic for naming Modules: https://github.com/chipsalliance/chisel/blob/8b60f1947d257ad8cdf3e5dcf72e459be868e18a/core/src/main/scala/chisel3/Module.scala#L507

I'm not going to sandbag this PR of this issue, but more careful handling either needs to be in this PR or come shortly after.

Reply:
We chatted offline but I think the best path forward is to just find a way to ban inner classes from using `HasAutoTypename` in the compiler plugin, then we don't need to worry about the `.getSimpleName` issue either because that only affects inner classes.